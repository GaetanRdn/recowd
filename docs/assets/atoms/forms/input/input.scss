@mixin rc-input($config) {
  $typographyConfig: map-get(map-get($config, typography), input);
  $colors: map-get($config, colors);
  $primary: map-get($colors, primary);

  .rc-input {
    appearance: none;
    outline: none;
    padding: 1rem 1.6rem;
    min-width: 18rem;
    border-top-left-radius: 0.4rem;
    border-top-right-radius: 0.4rem;
    border: none;
    margin: 0;

    @each $key in font-size, font-family, font-weight, line-height {
      #{$key}: map-get($typographyConfig, $key);
    }

    @each $key in color, background-color {
      #{$key}: map-get($primary, $key);
    }

    &:not(.rc-outlined) {
      border-bottom: 0.1rem solid map-get($primary, border-color);
    }

    &:not(.rc-outlined):focus {
      border-bottom-width: 0.2rem;
    }

    &:not(:disabled):hover {
      background-color: lighten(map-get($primary, background-color), 12%);
    }

    &:not(:disabled).rc-outlined {
      border-bottom-left-radius: 0.4rem;
      border-bottom-right-radius: 0.4rem;
      outline: 0.1rem solid map-get($primary, border-color);

      &:focus {
        outline-width: 0.2rem;
      }
    }

    &:disabled {
      border: none;
      color: darken(map-get($primary, color), 50%);
      border-bottom-left-radius: 0.4rem;
      border-bottom-right-radius: 0.4rem;
    }
  }
}
