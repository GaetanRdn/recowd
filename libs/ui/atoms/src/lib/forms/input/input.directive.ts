import {
  Directive,
  EventEmitter,
  HostBinding,
  HostListener,
  Input,
  NgModule,
  Optional,
  Output,
} from '@angular/core';
import {
  CoerceBoolean,
  CoerceString,
  Nullable,
  OnChangeFn,
  OnTouchedFn,
  TypedControlValueAccessor,
} from '@recowd/utility-types';
import { NgControl } from '@angular/forms';

@Directive({
  selector: 'input[rcInput]',
  host: {
    class: 'rc-input',
    '[class.rc-outlined]': 'outlined',
  },
})
export class InputDirective
  implements TypedControlValueAccessor<Nullable<string>>
{
  @Input() @CoerceBoolean() public outlined?: boolean;

  @HostBinding('value')
  @CoerceString()
  @Input()
  public value = '';

  @Output()
  public readonly valueChange: EventEmitter<string> = new EventEmitter<string>();

  constructor(@Optional() private readonly _ngControl?: NgControl) {
    if (_ngControl) {
      _ngControl.valueAccessor = this;
    }
  }

  private _disabled = false;

  get disabled(): boolean {
    return this._disabled || Boolean(this._ngControl?.disabled);
  }

  @Input()
  @CoerceBoolean()
  @HostBinding('disabled')
  set disabled(disabled: boolean) {
    this._disabled = disabled;
  }

  public writeValue(value: Nullable<string>): void {
    this.value = value ?? '';
  }

  public registerOnChange(fn: OnChangeFn<Nullable<string>>): void {
    this.onChange = fn;
  }

  public registerOnTouched(fn: OnTouchedFn): void {
    this.onTouched = fn;
  }

  @HostListener('input', ['$event.target.value'])
  private input(value: string): void {
    this.valueChange.emit(value);
    this.onTouched();
    this.onChange(value);
  }

  private onTouched: OnTouchedFn = () => {
    // default method if the directive is not used in a reactive form
  };

  private onChange: OnChangeFn<Nullable<string>> = () => {
    // default method if the directive is not used in a reactive form
  };
}

@NgModule({
  declarations: [InputDirective],
  exports: [InputDirective],
})
export class InputModule {}
