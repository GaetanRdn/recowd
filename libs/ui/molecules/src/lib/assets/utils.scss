@function map-get-deep($map, $level, $subLevel) {
  @return map-get($map, $level, $subLevel);
}

@mixin use-typography($typography) {
  @each $key in font-size, font-family, font-weight, line-height {
    #{$key}: map-get($typography, $key);
  }
}
