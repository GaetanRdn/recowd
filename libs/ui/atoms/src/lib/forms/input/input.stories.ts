import { Args, Meta, moduleMetadata, Story } from '@storybook/angular';
import { InputDirective } from './input.directive';
import { action } from '@storybook/addon-actions';
import {
  FormsModule,
  ReactiveFormsModule,
  UntypedFormControl,
} from '@angular/forms';
import { BADGE } from '@geometricpanda/storybook-addon-badges';

export default {
  title: 'atoms/forms/input',
  parameters: {
    badges: [BADGE.STABLE],
    jest: 'input.directive',
  },
  decorators: [
    moduleMetadata({
      imports: [InputDirective, ReactiveFormsModule, FormsModule],
    }),
  ],
  argTypes: {
    valueChange: { action: 'valueChanged' },
  },
} as Meta<InputDirective>;

const template: Story = (args: Args) => ({
  props: {
    disabled: args['disabled'] ?? false,
    valueChange: action('log'),
  },
  template: `
  <input rcInput placeholder="Type something"
         [disabled]="disabled" (valueChange)="valueChange($event)"/>`,
});

export const Default = template.bind({});
Default.args = {
  disabled: false,
};

const templateForTemplateDriven: Story = (args: Args) => ({
  props: {
    valueChange: action('log'),
    inputValue: '',
    disabled: args['disabled'],
  },
  template: `
  <p style="color: white; font-size: 16px;">inputValue: {{ inputValue }}</p>
  <input rcInput placeholder="Type something" [(ngModel)]="inputValue" (valueChange)="valueChange($event)"
  [disabled]="disabled" />`,
});

export const TemplateDriven = templateForTemplateDriven.bind({});
TemplateDriven.args = {
  disabled: false,
};

const templateForReactiveForms: Story = (args: Args) => ({
  props: {
    valueChange: action('log'),
    fc: new UntypedFormControl({
      value: args['initialValue'],
      disabled: args['disabledControl'],
    }),
  },
  template: `
  <p style="color: white; font-size: 16px;">inputValue: {{ fc.value }}</p>
  <input rcInput placeholder="Type something" [formControl]="fc" (valueChange)="valueChange($event)"/>`,
});

export const ReactiveForms = templateForReactiveForms.bind({});
ReactiveForms.args = {
  disabledControl: false,
  initialValue: '',
};
