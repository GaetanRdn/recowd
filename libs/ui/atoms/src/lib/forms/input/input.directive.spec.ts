import { Component } from '@angular/core';
import { TestBed, waitForAsync } from '@angular/core/testing';
import { InputDirective } from './input.directive';
import { TemplateLookup } from '@recowd/test/utils';
import { FormControl, ReactiveFormsModule } from '@angular/forms';

describe('InputDirective', () => {
  beforeEach(waitForAsync(() =>
    TestBed.configureTestingModule({
      declarations: [
        DefaultHostComponent,
        OutlinedHostComponent,
        DisabledHostComponent,
        ReactiveFormHostComponent,
        InputDirective,
      ],
      imports: [ReactiveFormsModule],
    }).compileComponents()));

  it('should create default', () => {
    // WHEN
    const templateLookup: TemplateLookup<DefaultHostComponent> =
      new TemplateLookup(DefaultHostComponent);
    templateLookup.detectChanges();

    // THEN
    expect(templateLookup.firstChildElement).toMatchSnapshot();
  });

  it('should emit valueChange on input', () => {
    // GIVEN
    const templateLookup: TemplateLookup<DefaultHostComponent> =
      new TemplateLookup(DefaultHostComponent);
    templateLookup.detectChanges();

    // WHEN
    templateLookup.triggerInput('input', 'test');
    templateLookup.detectChanges();

    // THEN
    expect(templateLookup.firstChildElement).toMatchSnapshot();
    expect(templateLookup.hostComponent.value).toEqual('test');
  });

  it('should create outlined', () => {
    // WHEN
    const templateLookup: TemplateLookup<OutlinedHostComponent> =
      new TemplateLookup(OutlinedHostComponent);
    templateLookup.detectChanges();

    // THEN
    expect(templateLookup.firstChildElement).toMatchSnapshot();
  });

  it('should create disabled', () => {
    // WHEN
    const templateLookup: TemplateLookup<DisabledHostComponent> =
      new TemplateLookup(DisabledHostComponent);
    templateLookup.detectChanges();

    // THEN
    expect(templateLookup.firstChildElement).toMatchSnapshot();
  });

  it('should create with form control', () => {
    // WHEN
    const templateLookup: TemplateLookup<ReactiveFormHostComponent> =
      new TemplateLookup(ReactiveFormHostComponent);
    templateLookup.detectChanges();

    // THEN
    expect(templateLookup.firstChildElement).toMatchSnapshot();
  });

  it('should set value with form control value', () => {
    // WHEN
    const templateLookup: TemplateLookup<ReactiveFormHostComponent> =
      new TemplateLookup(ReactiveFormHostComponent);
    templateLookup.hostComponent.fc.setValue('value');
    templateLookup.detectChanges();

    // THEN
    expect(templateLookup.firstChildElement).toMatchSnapshot();
    expect(
      templateLookup.getDirectiveOn('input', InputDirective).value
    ).toEqual('value');
  });

  it('should update control on input', () => {
    // GIVEN
    const templateLookup: TemplateLookup<ReactiveFormHostComponent> =
      new TemplateLookup(ReactiveFormHostComponent);
    templateLookup.detectChanges();

    // WHEN
    templateLookup.triggerInput('input', 'test');
    templateLookup.detectChanges();

    // THEN
    expect(templateLookup.firstChildElement).toMatchSnapshot();
    expect(templateLookup.hostComponent.fc.value).toEqual('test');
  });

  it('should be disabled when control is disabled', () => {
    // WHEN
    const templateLookup: TemplateLookup<ReactiveFormHostComponent> =
      new TemplateLookup(ReactiveFormHostComponent);
    templateLookup.hostComponent.fc.disable();
    templateLookup.detectChanges();

    // THEN
    expect(templateLookup.firstChildElement).toMatchSnapshot();
  });
});

@Component({
  template: `<input rcInput [(value)]="value" />`,
})
class DefaultHostComponent {
  public value?: string;
}

@Component({
  template: `<input rcInput outlined />`,
})
class OutlinedHostComponent {}

@Component({
  template: `<input rcInput disabled />`,
})
class DisabledHostComponent {}

@Component({
  template: `<input rcInput [formControl]="fc" />`,
})
class ReactiveFormHostComponent {
  public readonly fc: FormControl = new FormControl();
}
