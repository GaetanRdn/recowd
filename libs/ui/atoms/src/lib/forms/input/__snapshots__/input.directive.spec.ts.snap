// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputDirective should be disabled when control is disabled 1`] = `
<input
  class="rc-input ng-untouched ng-pristine"
  disabled=""
  rcinput=""
/>
`;

exports[`InputDirective should create default 1`] = `
<input
  class="rc-input"
  rcinput=""
/>
`;

exports[`InputDirective should create disabled 1`] = `
<input
  class="rc-input"
  disabled=""
  rcinput=""
/>
`;

exports[`InputDirective should create outlined 1`] = `
<input
  class="rc-input rc-outlined"
  outlined=""
  rcinput=""
/>
`;

exports[`InputDirective should create with form control 1`] = `
<input
  class="rc-input ng-untouched ng-pristine ng-valid"
  rcinput=""
/>
`;

exports[`InputDirective should emit valueChange on input 1`] = `
<input
  class="rc-input"
  rcinput=""
/>
`;

exports[`InputDirective should set value with form control value 1`] = `
<input
  class="rc-input ng-untouched ng-pristine ng-valid"
  rcinput=""
/>
`;

exports[`InputDirective should update control on input 1`] = `
<input
  class="rc-input ng-valid ng-touched ng-dirty"
  rcinput=""
/>
`;
